packages/date-picker/src/vaadin-date-picker-mixin.js

```diff
@@ -703,27 +702,57 @@ export const DatePickerMixin = (subclass) =>
       }
       if (!dateEquals(this[property], date)) {
         this[property] = date;
-        if (this.value) {
-          this.validate();
-        }
       }
     }
 
-    /** @private */
-    _valueChanged(value, oldValue) {
-      this._handleDateChange('_selectedDate', value, oldValue);
+    /**
+     * Override the value observer from `InputMixin` to implement custom
+     * handling of the `value` property. The date-picker doesn't forward
+     * the value directly to the input like the default implementation of `InputMixin`.
+     * Instead, it parses the value into a date, puts it in `_selectedDate` which
+     * is then displayed in the input with respect to the specified date format.
+     *
+     * @param {string | undefined} newValue
+     * @param {string | undefined} oldValue
+     * @protected
+     * @override
+     */
+    _valueChanged(newValue, oldValue) {
+      const oldSelectedDate = this._selectedDate;
+      this._handleDateChange('_selectedDate', newValue, oldValue);
+      const newSelectedDate = this._selectedDate;
+
+      this._toggleHasValue(!!newValue);
 
-      this._toggleHasValue(!!value);
+      if (oldSelectedDate !== newSelectedDate && newValue && oldValue !== undefined) {
```

Comment:
Isn't `oldSelectedDate !== newSelectedDate` always be truth-y? When `_selectedDate` is an instance of `Date`, then the comparison is always true. The other moment I found where `_selectedDate` is not a `Date` is when you change the value from something to empty value. On that case, `oldSelectedDate` is `null` and `newSelectedDate` is `''`. Am I missing something?

Reply:
> Isn't oldSelectedDate !== newSelectedDate always be truth-y?

There is a check before updating the selected date to avoid creating a new instance of Date if the date hasn't actually changed:

https://github.com/vaadin/web-components/blob/27062871c049b9adda1a49de64c38eee749f8ddb/packages/date-picker/src/vaadin-date-picker-mixin.js#L703-L705

Reply:
> The other moment I found where _selectedDate is not a Date is when you change the value from something to empty value. On that case, oldSelectedDate is null and newSelectedDate is ''

Not sure I understand what the case are we talking about here. Setting a new value via the `value` property? Or some other case?

Reply:
When you clear the value, then `_selectedDate` is an empty string.

Reply:
Rebased the PR.